File: src/main/java/no/nav/familie/ks/oppslag/tilgangskontroll/TilgangskontrollController.java
package no.nav.familie.ks.oppslag.tilgangskontroll;

import no.nav.familie.ks.oppslag.azure.AzureGraphService;
import no.nav.familie.ks.oppslag.azure.domene.Saksbehandler;
import no.nav.familie.ks.oppslag.personopplysning.PersonopplysningerService;
import no.nav.familie.ks.oppslag.personopplysning.domene.Personinfo;
import no.nav.familie.ks.oppslag.tilgangskontroll.domene.Tilgang;
import no.nav.security.token.support.core.api.ProtectedWithClaims;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.constraints.NotNull;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.http.ResponseEntity.ok;

@RestController
@RequestMapping(value = "/api/tilgang")
public class TilgangskontrollController {

    private AzureGraphService azureGraphService;
    private TilgangsKontrollService tilgangService;
    private PersonopplysningerService personService;

    @Autowired
    public TilgangskontrollController(AzureGraphService azureGraphService, TilgangsKontrollService tilgangsKontrollService, PersonopplysningerService personopplysningerService) {
        this.azureGraphService = azureGraphService;
        this.tilgangService = tilgangsKontrollService;
        this.personService = personopplysningerService;
    }

    @GetMapping(path = "/person")
    @ProtectedWithClaims(issuer = "azuread")
    public ResponseEntity tilgangTilPerson(@NotNull @RequestHeader(name = "Nav-Personident") String personIdent) {
        return sjekkTilgangTilBruker(personIdent);
    }

    private ResponseEntity sjekkTilgangTilBruker(String personIdent) {
        Saksbehandler saksbehandler = azureGraphService.getSaksbehandler();
        Personinfo personInfo = personService.hentPersoninfo(personIdent);
        Tilgang tilgang = tilgangService.sjekkTilgang(personIdent, saksbehandler, personInfo);
        return lagRespons(tilgang);
    }

    private ResponseEntity lagRespons(Tilgang tilgang) {
            return ok()
                    .contentType(APPLICATION_JSON)
                    .body(tilgang);
    }
}

File: src/main/java/no/nav/familie/integrasjoner/tilgangskontroll/TilgangskontrollController.kt
package no.nav.familie.integrasjoner.tilgangskontroll

import no.nav.familie.kontrakter.felles.PersonIdent
import no.nav.familie.kontrakter.felles.Tema
import no.nav.familie.kontrakter.felles.tilgangskontroll.Tilgang
import no.nav.security.token.support.core.api.ProtectedWithClaims
import org.springframework.context.annotation.Profile
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestHeader
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping(value = ["/api/tilgang"])
@Profile("!e2e")
class TilgangskontrollController(private val tilgangskontrollService: TilgangskontrollService) {
    @PostMapping(path = ["/v2/personer"])
    @ProtectedWithClaims(issuer = "azuread")
    fun tilgangTilPersoner(
        @RequestBody personIdenter: List<String>,
        @RequestHeader(name = "Nav-Tema") tema: Tema,
    ): List<Tilgang> {
        return tilgangskontrollService.sjekkTilgangTilBrukere(personIdenter, tema)
    }

    @PostMapping(path = ["/person-med-relasjoner"])
    @ProtectedWithClaims(issuer = "azuread")
    fun tilgangTilPersonMedRelasjoner(
        @RequestBody personIdent: PersonIdent,
        @RequestHeader(name = "Nav-Tema") tema: Tema,
    ): Tilgang {
        return tilgangskontrollService.sjekkTilgangTilPersonMedRelasjoner(personIdent.ident, tema)
    }
}
